modules/drivers/presto-jdbc/src/metabase/driver/presto_jdbc.clj

```diff
@@ -0,0 +1,470 @@
+  "Unwraps the C3P0 `pooled-conn` and returns the underlying `PrestoConnection` it holds."
+  [^C3P0ProxyConnection pooled-conn]
+  (.unwrap pooled-conn PrestoConnection))
+
+(defmethod sql-jdbc.execute/connection-with-timezone :presto-jdbc
+  [driver database ^String timezone-id]
+  ;; Presto supports setting the session timezone via a `PrestoConnection` instance method. Under the covers,
+  ;; this is equivalent to the `X-Presto-Time-Zone` header in the HTTP request (i.e. the `:presto` driver)
+  (let [conn            (.getConnection (sql-jdbc.execute/datasource database))
+        underlying-conn (pooled-conn->presto-conn conn)]
+    (try
+      (sql-jdbc.execute/set-best-transaction-level! driver conn)
+      (when-not (str/blank? timezone-id)
+        ;; set session time zone if defined
+        (.setTimeZoneId underlying-conn timezone-id))
+      (try
+        (.setReadOnly conn true)
+        (catch Throwable e
+          (log/debug e (trs "Error setting connection to read-only"))))
+      ;; as with statement and prepared-statement, cannot set holdability on the connection level
+      conn
+      (catch Throwable e
+        (.close conn)
+        (throw e)))))
+
+(defmethod sql-jdbc.execute/set-parameter [:presto-jdbc OffsetDateTime]
+  [_ prepared-statement i ^OffsetDateTime t]
+  ;; necessary because `PrestoPreparedStatement` does not support `OffsetDateTime` in its setObject method
+  ;; convert to a SQL `timestamp` at UTC (offset 0), and set that
+  (jdbc/set-parameter (t/sql-timestamp (t/with-offset-same-instant t (t/zone-offset 0))) prepared-statement i))
+
+(defmethod sql-jdbc.execute/set-parameter [:presto-jdbc ZonedDateTime]
+  [_ prepared-statement i ^OffsetDateTime t]
+  ;; necessary because `PrestoPreparedStatement` does not support `ZonedDateTime` in its setObject method
+  ;; convert to a SQL `timestamp` at UTC, and set that
+  (jdbc/set-parameter (t/sql-timestamp (t/with-zone-same-instant t (t/zone-id "UTC"))) prepared-statement i))
+
+(defn- date-time->substitution [ts-str]
+  (sql.params.substitution/make-stmt-subs "from_iso8601_timestamp(?)" [ts-str]))
+
+(defmethod sql.params.substitution/->prepared-substitution [:presto-jdbc ZonedDateTime]
+  [_ ^ZonedDateTime t]
+  ;; for native query parameter substitution, in order to not conflict with the `PrestoConnection` session time zone
+  ;; (which was set via report time zone), it is necessary to use the `from_iso8601_timestamp` function on the string
+  ;; representation of the `ZonedDateTime` instance, but converted to the report time zone
+  (let [report-zone       (qp.timezone/report-timezone-id-if-supported :presto-jdbc)
+        ^ZonedDateTime ts (if (str/blank? report-zone) t (t/with-zone-same-instant t (t/zone-id report-zone)))]
+    ;; the `from_iso8601_timestamp` only accepts timestamps with an offset (not a zone ID), so only format with offset
+    (date-time->substitution (.format ts DateTimeFormatter/ISO_OFFSET_DATE_TIME))))
```

Comment:
You can probably use `(u.date/format t)` to do this if you don't want to deal with the Java methods. Doesn't really make a big difference tho I guess

Are you sure

Reply:
It seems like having this output `timestamp '2021-06-23 14:09:00 -07:00'` or whatever the actual Presto literal format is might be preferable to `from_iso8601_timestamp(?)` but I guess it ultimately doesn't make a difference

Reply:
Yeah I changed these to use the first form (which is actually represented as a `CAST` in order to work with prepared statements/String param value).